feat: add `lapack/base/dlange` (#7195)
stdlib-js/stdlib: lib/node_modules/@stdlib/lapack/base/dlange/README.md

<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dlange

> > LAPACK routine to compute the value of the one norm, or the frobenius norm, or the infinity norm, or the element with the largest absolute value of a real matrix `A`.

Reviewer: ```suggestion
> LAPACK routine to compute the value of the one norm, the frobenius norm, the infinity norm, or the element with the largest absolute value of a real matrix `A`.
```

Applies here and everywhere else in this PR.

<section class="intro">

> The `dlange` routine computes the value of a specified norm of a real M-by-N matrix `A`. The norm to be computed is selected using the parameter `norm`, which may specify the **Max norm**, **One norm**, **Infinity norm**, or **Frobenius norm**.

Reviewer: ```suggestion
This routine computes the value of a specified norm of a real M-by-N matrix `A`.
```

<!-- <equation class="equation" label="eq:frobenius_norm" align="center" raw="\|A\|_F = \left(\sum_{i=1}^M \sum_{j=1}^N |a_{i,j}|^2 \right)^{1/2}" alt="Definition of frobenius norm of a matrix."> -->

```math
\|A\|_F = \left(\sum_{i=1}^M \sum_{j=1}^N |a_{i,j}|^2 \right)^{1/2}
```

<!-- </equation> -->

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var dlange = require( '@stdlib/lapack/base/dlange' );
```

#### dlange( norm, M, N, A, LDA, work )

Computes the value of the one norm, or the frobenius norm, or the infinity norm, or the element with the largest absolute value of a real matrix `A`.

<!-- eslint-disable max-len -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var A = new Float64Array( [ 1.0, 4.0, 7.0, 10.0, 2.0, 5.0, 8.0, 11.0, 3.0, 6.0, 9.0, 12.0 ] );

/*
    A = [
        [ 1.0, 4.0, 7.0, 10.0 ],
        [ 2.0, 5.0, 8.0, 11.0 ],
        [ 3.0, 6.0, 9.0, 12.0 ]
    ]
*/

var work = new Float64Array( 3 );
var out = dlange( 'row-major', 'frobenius', 3, 4, A, 4, work );
// returns ~25.5
```

The function has the following parameters:

-   **order**: storage layout.

> -   **norm**: specifies the type of norm to be calculated, should be one of the following: `max`, `one`, `frobenius`, or `infinity`.

Reviewer: ```suggestion
-   **norm**: specifies the type of norm to be calculated, should be one of the following: `'max'`, `'one'`, `'frobenius'`, or `'infinity'`.
```

For string literals, always enclose in single quotes. Otherwise, ambiguous the value type. Here, `infinity` could mean `'infinity'` or the numeric literal `infinity`. Enclosing in quotes makes this explicit.

        [ 2.0, 5.0, 8.0, 11.0 ],
        [ 3.0, 6.0, 9.0, 12.0 ]
    ]
*/

// Create offset views...
var A = new Float64Array( A0.buffer, A0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
var work = new Float64Array( work0.buffer, work0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

var out = dlange( 'row-major', 'frobenius', 3, 4, A, 4, work );
// returns ~25.5
```

<!-- lint disable maximum-heading-length -->

#### dlange.ndarray( norm, A, strideA1, strideA2, offsetA, work, strideWork, offsetWork )

Computes the value of the one norm, or the frobenius norm, or the infinity norm, or the element with the largest absolute value of a real matrix `A` using alternative indexing semantics.

<!-- eslint-disable max-len -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var A = new Float64Array( [ 1.0, 4.0, 7.0, 10.0, 2.0, 5.0, 8.0, 11.0, 3.0, 6.0, 9.0, 12.0 ] );

/*
    A = [
        [ 1.0, 4.0, 7.0, 10.0 ],
        [ 2.0, 5.0, 8.0, 11.0 ],
        [ 3.0, 6.0, 9.0, 12.0 ]
    ]
*/

var work = new Float64Array( 3 );
var out = dlange.ndarray( 'frobenius', 3, 4, A, 4, 1, 0, work, 1, 0 );
// returns ~25.5
```

The function has the following additional parameters:

> -   **norm**: specifies the type of norm to be calculated, should be one of the following: `max`, `one`, `frobenius`, or `infinity`.

Reviewer: Same comment.

<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# dlange

> LAPACK routine to compute the value of the one norm, or the frobenius norm, or the infinity norm, or the element with the largest absolute value of a real matrix `A`.

<section class="intro">

The `dlange` routine computes the value of a specified norm of a real M-by-N matrix `A`. The norm to be computed is selected using the parameter `norm`, which may specify the **Max norm**, **One norm**, **Infinity norm**, or **Frobenius norm**.

The supported norms are:

-   **Max Absolute Value** (`norm` = `'max'`): returns the largest absolute element in `A`.

> <!-- <equation class="equation" label="eq:lu_decomposition" align="center" raw="\|A\|_{\max} = \max_{i,j} |a_{i,j}|" alt="Maximum absolute value of a matrix."> -->

Reviewer: @aayush0325 I think you need to update the label for this and the other equations. Each label should be unique. This appears to be a copy-paste mistake.

<!-- <equation class="equation" label="eq:lu_decomposition" align="center" raw="\|A\|_F = \left(\sum_{i=1}^M \sum_{j=1}^N |a_{i,j}|^2 \right)^{1/2}" alt="Definition of frobenius norm of a matrix."> -->

```math
\|A\|_F = \left(\sum_{i=1}^M \sum_{j=1}^N |a_{i,j}|^2 \right)^{1/2}
```

<!-- </equation> -->

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var dlange = require( '@stdlib/lapack/base/dlange' );
```

#### dlange( norm, M, N, A, LDA, work )

Computes the value of the one norm, or the frobenius norm, or the infinity norm, or the element with the largest absolute value of a real matrix `A`.

<!-- eslint-disable max-len -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var A = new Float64Array( [ 1.0, 4.0, 7.0, 10.0, 2.0, 5.0, 8.0, 11.0, 3.0, 6.0, 9.0, 12.0 ] );

/*
    A = [
        [ 1.0, 4.0, 7.0, 10.0 ],
        [ 2.0, 5.0, 8.0, 11.0 ],
        [ 3.0, 6.0, 9.0, 12.0 ]
    ]
*/

var work = new Float64Array( 3 );
var out = dlange( 'row-major', 'frobenius', 3, 4, A, 4, work );
// returns ~25.5
```

The function has the following parameters:

-   **order**: storage layout.

> -   **norm**: specifies the type of norm to be calculated, should be one of the following: `max`, `one`, `frobenius` or `infinity`.

Reviewer: ```suggestion
-   **norm**: specifies the type of norm to be calculated, should be one of the following: `max`, `one`, `frobenius`, or `infinity`.
```

Oxford comma.

-   **M**: number of rows in `A`.
-   **N**: number of columns in `A`.
-   **A**: input [`Float64Array`][mdn-float64array].
-   **LDA**: stride of the first dimension of `A` (a.k.a., leading dimension of the matrix `A`).
-   **work**: [`Float64Array`][mdn-float64array] used as a temporary workspace.

> `work` should have `N` indexed elements if calculating the one norm in a row-major layout and `M` indexed elements if calculating the infinity norm in column-major layout, in all other cases it is fine to pass a dummy [`Float64Array`][mdn-float64array].

Reviewer: ```suggestion
-   **work**: [`Float64Array`][mdn-float64array] used as a temporary workspace. When `A` is row-major and `norm` is `one`, `work` should have at least `N` indexed elements. When `A` is column-major` and `norm` is `infinity`, work` should have at least `M` indexed elements. In all other cases, `work` is not used and can have any number of elements, including zero.
```

#### dlange.ndarray( norm, A, strideA1, strideA2, offsetA, work, strideWork, offsetWork )

Computes the value of the one norm, or the frobenius norm, or the infinity norm, or the element with the largest absolute value of a real matrix `A` using alternative indexing semantics.

<!-- eslint-disable max-len -->

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var A = new Float64Array( [ 1.0, 4.0, 7.0, 10.0, 2.0, 5.0, 8.0, 11.0, 3.0, 6.0, 9.0, 12.0 ] );

/*
    A = [
        [ 1.0, 4.0, 7.0, 10.0 ],
        [ 2.0, 5.0, 8.0, 11.0 ],
        [ 3.0, 6.0, 9.0, 12.0 ]
    ]
*/

var work = new Float64Array( 3 );
var out = dlange.ndarray( 'frobenius', 3, 4, A, 4, 1, 0, work, 1, 0 );
// returns ~25.5
```

The function has the following additional parameters:

-   **norm**: specifies the type of norm to be calculated, should be one of the following: `max`, `one`, `frobenius` or `infinity`.
-   **M**: number of rows in `A`.
-   **N**: number of columns in `A`.
-   **A**: input [`Float64Array`][mdn-float64array].
-   **strideA1**: stride of the first dimension of `A`.
-   **strideA2**: stride of the second dimension of `A`.
-   **offsetA**: starting index for `A`.

> -   **work**: [`Float64Array`][mdn-float64array] used as a temporary workspace

Reviewer: Same changes as above.
